api/src/main/java/org/apache/iceberg/RewriteFiles.java

```diff
@@ -97,6 +97,22 @@ default RewriteFiles addFile(DeleteFile deleteFile) {
         this.getClass().getName() + " does not implement addFile");
   }
 
+  /**
+   * Add a new delete file with the given data sequence number.
+   *
+   * <p>This rewrite operation may change the size or layout of the delete files. When applicable,
+   * it is also recommended to discard delete records for files that are no longer part of the table
+   * state. However, the set of applicable delete records must never change.
+   *
+   * @param deleteFile a new delete file
+   * @param dataSequenceNumber data sequence number to append on the file
+   * @return this for method chaining
+   */
+  default RewriteFiles addFile(DeleteFile deleteFile, long dataSequenceNumber) {
```

Comment:
My original thought was to use `DeleteFile.dataSequenceNumber()` if set (could be done by extending the builder) but this approach can be a bit more reliable as it is explicit. Question: is there any case when the *added* new delete file data sequence number can be wrong?

Any thoughts, @rdblue @RussellSpitzer?

Reply:
If we consume it directly from the file, it would mean this would apply for appends too.

Reply:
Right now, we don't have a way to set the sequence number in the public API. I think that's a good thing because we don't want people to think that this is something they can generally control. So I do prefer setting this here.

However, for the implementation in `BaseRewriteDataFiles` I think it would make sense to set the sequence number and avoid needing the `DeleteFileHolder`. Not a big deal either way, though.

Reply:
> Right now, we don't have a way to set the sequence number in the public API.

That's can be easily changed by extending `TableMetadata$Builder`. I am not a big fan of `DeleteFileHolder` and I do think places that actually set data sequence numbers (like rewrite position deletes) would benefit from passing it as part of delete file object. That said, I am concerned about edge cases when the passed object may contain a wrong data sequence number. Are there use cases we can think of? Like cherry-picks maybe?

Reply:
Yeah, cherry picking seems like a case where we'd have the wrong sequence number. The good news is that we don't allow cherry picking unless the commit is an append or a dynamic partition overwrite. Otherwise we lose too much information.

I'm also a bit concerned about letting sequence number on the file be used. I'd probably opt to ignore it and have an explicit sequence number in the API like this.

Reply:
Okay, let's stick with a safer option. Let me review the rest of the PR.